From 35af259500eacd23447775313097720bbd8d4fc3 Mon Sep 17 00:00:00 2001
From: Juan Gutierrez <juan.gutierrez@nxp.com>
Date: Wed, 20 Apr 2016 14:38:02 -0500
Subject: [PATCH 12/23] rpmsg: imx: support for configurable vring address by
 device tree

vring memory address was hardcoded at the top of the 1GB RAM.
For systems with a memory map with less or different than 1GB,
the hardcoded value might be not correct and cause issues.

This patch add the support to pass the vring address from device
tree configuration on the reg platfrom argument in the follwing
format:

    reg = <vring_address vring_size>

For example, for a 512MB system, with the rpmgs vring placed at
top of the memory the configuration will look like below:

    &rpmsg{
           reg = <0x9FFF0000 0x8000>;
    };

Signed-off-by: Juan Gutierrez <juan.gutierrez@nxp.com>
---
 arch/arm/mach-imx/imx_rpmsg.c | 17 ++++++++++++++---
 1 file changed, 14 insertions(+), 3 deletions(-)

diff --git a/arch/arm/mach-imx/imx_rpmsg.c b/arch/arm/mach-imx/imx_rpmsg.c
index 7ccaf67..45eae65 100644
--- a/arch/arm/mach-imx/imx_rpmsg.c
+++ b/arch/arm/mach-imx/imx_rpmsg.c
@@ -281,6 +281,8 @@ static int imx_rpmsg_probe(struct platform_device *pdev)
 {
 	int i, ret = 0;
 	struct device_node *np = pdev->dev.of_node;
+	struct resource *res;
+	resource_size_t size;
 
 	for (i = 0; i < ARRAY_SIZE(imx_rpmsg_vprocs); i++) {
 		struct imx_rpmsg_vproc *rpdev = &imx_rpmsg_vprocs[i];
@@ -289,9 +291,18 @@ static int imx_rpmsg_probe(struct platform_device *pdev)
 			ret = of_device_is_compatible(np, "fsl,imx7d-rpmsg");
 			ret |= of_device_is_compatible(np, "fsl,imx6sx-rpmsg");
 			if (ret) {
-				/* hardcodes here now. */
-				rpdev->vring[0] = 0xBFFF0000;
-				rpdev->vring[1] = 0xBFFF8000;
+
+				res  = platform_get_resource(pdev, IORESOURCE_MEM, 0);
+
+				if (res) {
+					size = resource_size(res);
+					rpdev->vring[0] = res->start;
+					rpdev->vring[1] = res->start + size;
+				} else {
+					/* hardcodes here now. */
+					rpdev->vring[0] = 0xBFFF0000;
+					rpdev->vring[1] = 0xBFFF8000;
+				}
 			}
 		} else {
 			break;
-- 
1.9.1

